server/src/main/java/com/cloud/user/PasswordPolicyImpl.java

```diff
@@ -0,0 +1,245 @@
+    }
+
+    protected void validateIfPasswordContainsTheUsername(String password, String username, Long domainId) {
+        logger.trace(String.format("Validating if the new password for user [%s] contains their username.", username));
+
+        if (getPasswordPolicyAllowPasswordToContainUsername(domainId)) {
+            logger.trace(String.format("Allow password to contain username is true; therefore, we will not validate if the password contains the username of user [%s].", username));
+            return;
+        }
+
+        if (StringUtils.containsIgnoreCase(password, username)) {
+            logger.error(String.format("User [%s] informed a new password that contains their username; however, the this is not allowed as configured in [%s]. "
+                    + "Refusing the user's new password.", username, PasswordPolicyAllowPasswordToContainUsername.key()));
+            throw new InvalidParameterValueException("User password should not contain their username.");
+        }
+
+        logger.trace(String.format("The new password for user [%s] complies with the policy of allowing passwords to contain username [%s].", username,
+                PasswordPolicyAllowPasswordToContainUsername.key()));
+    }
+
+    protected void validateIfPasswordMatchesRegex(String password, String username, Long domainId) {
+        String passwordPolicyRegex = getPasswordPolicyRegex(domainId);
+
+        logger.trace(String.format("Validating if the new password for user [%s] matches regex [%s] defined in the configuration [%s].",
+                username, passwordPolicyRegex, PasswordPolicyRegex.key()));
+
+        if (passwordPolicyRegex == null){
+            logger.trace(String.format("Regex is null; therefore, we will not validate if the new password matches with regex for user [%s].", username));
+            return;
+        }
+
+        if (!password.matches(passwordPolicyRegex)){
+            logger.error(String.format("User [%s] informed a new password that does not match with regex [%s]. Refusing the user's new password.", username, passwordPolicyRegex));
+            throw new InvalidParameterValueException("User password does not match with password policy regex.");
+        }
+
+        logger.trace(String.format("The new password for user [%s] complies with the policy of matching regex [%s].", username,
+                PasswordPolicyRegex.key()));
+    }
+
+    @Override
+    public String getConfigComponentName() {
+        return PasswordPolicyImpl.class.getSimpleName();
+    }
+
+    @Override
+    public ConfigKey<?>[] getConfigKeys() {
+        return new ConfigKey<?>[]{PasswordPolicyMinimumLength, PasswordPolicyMinimumSpecialCharacters, PasswordPolicyMinimumUppercaseLetters, PasswordPolicyMinimumLowercaseLetters,
+                PasswordPolicyMinimumDigits, PasswordPolicyAllowPasswordToContainUsername, PasswordPolicyRegex
+        };
+    }
+
+    public Integer getPasswordPolicyMinimumLength(Long domainId) {
```

Comment:
These methods may not be required, we can directly get the value from the configuration and it is as readable as these methods.

Reply:
Although this is true, using these methods makes unit testing easier and the code more organized.